### DISCUSSION TITLE
Translations in API

### DISCUSSION THREAD

**Discussion by alexcroox** [s00d/nuxt-i18n-micro#66] [Q&A] [answered]
Considering a swith from @nuxtjs/i18n.

I need to also translate API responses but couldn't find anything about this in the docs.

https://i18n.nuxtjs.org/docs/guide/server-side-translations

Would $t work out of the box in the server/api folder?

**Comment by s00d** [comment]
Hello! Currently, this functionality isn't available. I'm not entirely sure how it could be utilized or why it's needed in this context. If you could provide an example of how you're trying to use it and explain your specific requirements, I'd be happy to consider what might need to be implemented.

**Comment by alexcroox** [comment]
I'd like to return fully translated error messages from my API endpoints in Nuxt. Whether that's in response to the client making a request, or by third party clients utilising my API. If I'm displaying errors to the user I'd like them to be in their preferred language, and I'm not always in control of the client to be just sending back translation paths.

Here is an example of sending back a translated error using Nuxt i18n server side translations:

![SCR-20241202-lgyk](https://github.com/user-attachments/assets/8fb4e2be-d65c-4282-9417-b208db258124)

See their docs for more context about how it works: https://i18n.nuxtjs.org/docs/guide/server-side-translations

Thanks

**Comment by s00d** [comment]
You can simply create an error page that dynamically displays the error description based on the user's locale and the error code. At the moment, I see this task as overly complex and not particularly necessary.

---

Regarding the implementation of this functionality in **nuxt-i18n**: it seems that for every server request (and possibly any kind of request—this needs testing), a new Nuxt instance is created, and all translations are stored in memory. This behavior can be observed in the code:  

- [Line 16](https://github.com/nuxt-modules/i18n/blob/ea7aa14a6d6ba07f40a5c7e7795f5c247530651f/src/runtime/server/plugin.ts#L16)
- [Line 24](https://github.com/nuxt-modules/i18n/blob/ea7aa14a6d6ba07f40a5c7e7795f5c247530651f/src/runtime/server/plugin.ts#L24)

This explains why the application consumes so much RAM and experiences memory leaks. Such an implementation, created to support unnecessary functionality, results in serious performance issues.  

I would not recommend using this functionality, as it can cause significant performance problems and memory management issues in production environments.

**Comment by alexcroox** [comment]
Fair enough, I’ll stick with Nuxt i18n for now then as I have third party users that use my API and I don’t have to provide a large error code glossary for them currently. Also I have native apps that don’t update that often so adding new error codes on the backend only wouldn’t work. 

Good to know about potential performance issues though thanks!

**Comment by s00d** [comment, marked as answer]
I added [Server-Side Translations](https://s00d.github.io/nuxt-i18n-micro/guide/server-side-translations) in version 1.36.1. I should mention right away that the implementation is quite basic, but it will keep translation files in memory. Additionally, on the server, only translations from the root folder are used; subfolders are ignored.

[example](https://github.com/s00d/nuxt-i18n-micro/blob/main/playground/server/api/server.js)

**Comment by alexcroox** [comment]
Wow thank you so much it looks great! I'll give a migration to nuxt-i18n-micro a go tomorrow!

~I use Cloudflare pages/workers for my API so I assume the in-memory cache won't apply too often for me (serverless requests where the server isn't held in memory) but it's not too much of a penalty to load the json for each request is it? Not sure there's a way around that anyway unless it's in a global variable (they get re-used between requests on the same isolate)~

Edit: just seen that the cache is in a global variable so that will be re-used when it can in serverless environments, nice! 👌
https://github.com/s00d/nuxt-i18n-micro/blob/main/src/runtime/translationHelper.ts#L3

**Reply by s00d** [reply]
I’m using a global variable because I couldn’t think of another way. It’s not a big problem, but we need to remember it’s there and keep the translation files small.

**Reply by alexcroox** [reply]
Yeah the global variable is exactly what we need for cache re-use in serverless environments so thank you!!
